src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs

```diff
@@ -0,0 +1,275 @@
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByApplicationPolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByCertificatePolicy:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByCertificatePolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByExtension:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.ExtensionOrAttribute);
+                        findPal.FindByExtension(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByKeyUsage:
+                    {
+                        X509KeyUsageFlags keyUsage = ConfirmedX509KeyUsage(findValue);
+                        findPal.FindByKeyUsage(keyUsage);
+                        break;
+                    }
+                    case X509FindType.FindBySubjectKeyIdentifier:
+                    {
+                        byte[] keyIdentifier = ConfirmedCast<string>(findValue).DecodeHexString();
+                        findPal.FindBySubjectKeyIdentifier(keyIdentifier);
+                        break;
+                    }
+                    default:
+                        throw new CryptographicException(SR.Cryptography_X509_InvalidFindType);
+                }
+            }
+           
+            return results;
+        }
+
+        private static T ConfirmedCast<T>(object findValue)
+        {
+            Debug.Assert(findValue != null);
+
+            if (findValue.GetType() != typeof(T))
+                throw new CryptographicException(SR.Cryptography_X509_InvalidFindValue);
+
+            return (T)findValue;
+        }
+
+        private static string ConfirmedOidValue(IFindPal findPal, object findValue, OidGroup oidGroup)
+        {
+            string maybeOid = ConfirmedCast<string>(findValue);
+
+            if (maybeOid.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidOidValue);
+            }
+
+            return findPal.NormalizeOid(maybeOid, oidGroup);
+        }
+
+        private static X509KeyUsageFlags ConfirmedX509KeyUsage(object findValue)
+        {
+            if (findValue is X509KeyUsageFlags)
+                return (X509KeyUsageFlags)findValue;
+
+            if (findValue is int)
+                return (X509KeyUsageFlags)(int)findValue;
+
+            if (findValue is uint)
+                return (X509KeyUsageFlags)(uint)findValue;
+
+            string findValueString = findValue as string;
+
+            if (findValueString != null)
+            {
+                if (findValueString.Equals("DigitalSignature", StringComparison.OrdinalIgnoreCase))
```

Comment:
Could these mappings be in a static readonly Dictionary for faster lookups? (I suspect the Dictionary lookup would be faster than all the if checks, especially for strings towards the end of the if checks, but haven't measured to confirm).